### ISSUE TITLE
Kubemark-5k cluster failing performance tests

### ISSUE THREAD

**Issue by shyamjvs** [kubernetes/kubernetes#63025]
Our kubemark-5k job (https://k8s-testgrid.appspot.com/sig-scalability-kubemark#kubemark-5000) started failing sometime mid- last week. E.g failed runs:
- https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/logs/ci-kubernetes-kubemark-gce-scale/963
- https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/logs/ci-kubernetes-kubemark-gce-scale/964
- https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/logs/ci-kubernetes-kubemark-gce-scale/965

From run-964's build-log for e.g, I'm finding that quite some pods weren't started by hollow-nodes during the load test:

```
I0419 22:36:53.886] I0419 22:36:53.882724   15900 runners.go:176] Pod load-medium-1232-9bpsg    hollow-node-kmznc       Running <nil>
I0419 22:36:53.886] I0419 22:36:53.882733   15900 runners.go:176] Pod load-medium-1232-9stph    hollow-node-2lq5z       Pending <nil>
I0419 22:36:53.887] I0419 22:36:53.882742   15900 runners.go:176] Pod load-medium-1232-btvw9    hollow-node-7bqx9       Running <nil>
...
I0419 22:36:53.892] Apr 19 22:36:53.882: INFO: Unexpected error occurred: Only 29 pods started out of 30
```

My suspicion is that some hollow-nodes are dying due to resource starvation and hence not able to run pods that are assigned to those.
It's quite likely related to @krzysied 's change adding kube-dns to kubemark (which I think merged around that time).

cc @kubernetes/sig-scalability-bugs @wojtek-t

**Comment by shyamjvs** [comment]
/assign @krzysied 
Krzysiek - could you please take a look?

**Comment by wojtek-t** [comment]
My very strong suspicion is that ti's actually due to:
https://github.com/kubernetes/test-infra/pull/7751/files

**Comment by shyamjvs** [comment]
That's a good point - I missed that. I also feel it's more likely that one (also because density test is mostly passing).
I guess we have (at least) couple options now:
- revert that change
- increase no. of real nodes for more resources

Wdyt?

**Comment by wojtek-t** [comment]
It depends how many many more real nodes we will need.
If the difference is relatively small, I would vote for that. If not, maybe revert it...

**Comment by shyamjvs** [comment]
I suggest that we:

- revert the change and verify if the job passes
- if it's green, then try to estimate how much more resources we'd need (and decide based on it whether to increase)
- if it's red, find what else may be the cause

@krzysied can carry on the above if that's fine.

**Comment by wojtek-t** [comment]
sgtm

**Comment by krzysied** [comment]
OK. I'm reverting the changes.
https://github.com/kubernetes/test-infra/pull/7842

**Comment by shyamjvs** [comment]
Seems like the revert made the job green - https://k8s-testgrid.appspot.com/sig-scalability-kubemark#kubemark-5000
So now that we've confirmed it's due to proxier change, I believe we can continue:

- estimating how much more resources we'd need to run with proxier enabled
- verify if there's any other issue besides just shortage of resources

**Comment by wojtek-t** [comment]
> estimating how much more resources we'd need to run with proxier enabled

+1
As an experiment I would try bumping say by 2x or 3x and look at usage. That should give us estimation what we really need.

> verify if there's any other issue besides just shortage of resources

Proxier is used in kube-proxy on real clusters. Since we already enabled services in real clusters and don't observe other problems, I'm highly convinced that this is resource problem.

**Comment by shyamjvs** [comment]
> Proxier is used in kube-proxy on real clusters. Since we already enabled services in real clusters and don't observe other problems, I'm highly convinced that this is resource problem.

That's true. Though I feel it might be better to confirm - as there may be some problem for e.g with our kube-proxy mock.
That said, estimating resources is likely the more important part.

**Comment by krzysied** [comment]
I run the test with increased resources x2. The test failed - still looks like insufficient memory.
I used 60GB nodes (normally we use 30GB) and it didn't solve the problem. Do we want to try with x3 resources?

Btw, without real proxier the test executed fine.

**Comment by wojtek-t** [comment]
We definitely want to know how many resources we would need.
I can imagine ways to enable it with some tuning (e.g. instead of updating iptables every 1s or whatever the default is, do this every 10s). But we need to know what the overhead of it is before we decide to tune something.

**Comment by krzysied** [comment]
Ok. I will try to run it with x3 resources.

**Comment by krzysied** [comment]
Good news - x3 passed the test! Now we are sure that this is only resource related thing.
I used 123 nodes, 60GB each.

**Comment by wojtek-t** [comment]
Thanks for running that.

Now, looking in to logs of hollow-proxies, I'm seeing this (though on 100-node kubemark):
```
I0511 05:41:27.562448       7 proxier.go:346] minSyncPeriod: 5s, syncPeriod: 30s, burstSyncs: 2
```

[BTW - I don't really know where 5s is coming from - 30s is coming from default:
https://github.com/kubernetes/kubernetes/blob/master/pkg/proxy/apis/kubeproxyconfig/v1alpha1/defaults.go#L56
It would be great if you could figure this out].


The reason I ask for that is that now (when we know it's a matter of resources), we can try reducing the frequency of those resyncs in kube-proxy and that should help.

So let`s for example try the configuration where:
- minSyncPeriod is 30s
- syncPeriod is 120s

and get back to say 1.25x resources (we will need a bit more than we needed, and I don't know how much slack we have now).

My hope is that it should be enough and we would be able to enable real proxier in tests in that case.

**Comment by wojtek-t** [comment]
@krzysied - i think in the past you told me offline that it's mostly due to memory. I started thinking if that's really the case (based on memory usage of kube-proxies in real clusters).

When I think more about it, I actually started thinking that maybe what's happening is:
- hollow-proxy is cpu-starved
- because of that it's not able to process things fast enough, we are accumulating a bunch of incoming changes
- they are resulting in increasing memory usage

So I think that we should actually try bumping that minSyncPeriod and syncPeriod significantly and try to check if that would be enough to enable proxier.

**Comment by wojtek-t** [comment]
We can also try to profile kube-proxy to see where exactly it's spending time now.

**Comment by wojtek-t** [comment]
Actually, profiles are relatively obvious.
[profile001.pdf](https://github.com/kubernetes/kubernetes/files/2166931/profile001.pdf)

Optimizing ReadLine would give us huge benefits in terms of both cpu-usage and memory allocations.
https://github.com/kubernetes/kubernetes/pull/65815/ is small step towards that

**Comment by wojtek-t** [comment]
> [BTW - I don't really know where 5s is coming from - 30s is coming from default:
https://github.com/kubernetes/kubernetes/blob/master/pkg/proxy/apis/kubeproxyconfig/v1alpha1/defaults.go#L56
It would be great if you could figure this out].

So now I finally know where it is coming from. It's hardcoded in hollow_proxy.go:
https://github.com/kubernetes/kubernetes/blob/master/pkg/kubemark/hollow_proxy.go#L86

So i will make it configurable so that we can experiment with this a bit.
